Add BarChart component

samshara opened the pull request.

samshara committed:
Add BarChart component

tnagorra reviewed (commented):
I am not sure we are merging this to toggle-ui. Do we need this component?

tnagorra commented on src/components/BarChart/index.tsx:
```diff
@@ -0,0 +1,102 @@
+        },
+    }  = props;
+
+    const svgRef = useRef(null);
```
Type should be defined for svgRef

tnagorra commented on src/components/BarChart/index.tsx:
```diff
@@ -0,0 +1,102 @@
+            .attr('y', (d: K) => y(valueSelector(d)))
+            .attr('height', (d: K) => y(0) - y(valueSelector(d)))
+            .attr('width', x.bandwidth());
+    }, [data]);
```
Not all dependencies are listed here!

frozenhelium replied:
This is not the d4 method. @tnagorra should we use d4 method or this looks good?

tnagorra replied:
I don't think this is how we should write them!
Can you explain a bit more on the d4 method though?

tnagorra replied:
Without select, selectAll, enter, append, data, attr, style?

tnagorra replied:
@samshara for reference, https://d4.js.org/

tnagorra commented on src/components/BarChart/index.tsx:
```diff
@@ -0,0 +1,102 @@
+            .attr('width', width)
+            .attr('height', height);
+
+        const x = scaleBand()
+            .domain(data.map(d => labelSelector(d)))
+            .range([left, width - right])
+            .padding(0.1);
+
+        const y = scaleLinear()
+            .domain([0, max(data, valueSelector)])
+            .nice()
+            .range([height - bottom, top]);
+
+        const xAxis = axisBottom(x);
+
+        const yAxis = axisLeft(y);
```
can we not just memoize these?

tnagorra replied:
i don't know if it's better to use separate useEffect hooks for different side-effects (d3 invocation)

samshara committed:
Use react to render all svg elements

 d3.js is used for calculation only

tnagorra commented on src/components/BarChart/index.tsx:
```diff
@@ -38,61 +35,108 @@ function BarChart<K>(props: Props<K>) {
+        .range([height - bottom, top]),
+    [data, valueSelector, bottom, top, height]);
+
+    const bars = useMemo(() => data.map((d) => (
```
Maybe we don't need to memoize the jsx outputs?

samshara committed:
Remove memoized jxs

tnagorra committed:
Add a namespace Visuzations for Barchart

- Add typings for d3 dependencies
- Add depdendency d3-array
- Fix typing issues

frozenhelium commented on src/components/BarChart/index.tsx:
```diff
@@ -0,0 +1,152 @@
+import React, { useMemo } from 'react';
+
+import { scaleBand, scaleLinear } from 'd3-scale';
+import { max } from 'd3-array';
```
@tnagorra we should use `max` from fujs right?

frozenhelium commented on package.json:
```diff
@@ -42,6 +42,10 @@
         "@babel/runtime-corejs3": "^7.11.2",
         "@storybook/client-api": "^6.0.21",
         "@togglecorp/fujs": "^1.9.2-beta.0",
+        "d3-array": "^2.8.0",
```
is this required?

frozenhelium commented on package.json:
```diff
@@ -42,6 +42,10 @@
+        "d3-array": "^2.8.0",
+        "d3-axis": "^2.0.0",
+        "d3-scale": "^3.2.3",
+        "d3-selection": "^2.0.0",
```
is this required?

frozenhelium committed:
Improve barchart component
